src/systems/tips.js

```diff
@@ -0,0 +1,274 @@
+    if (!userinput.activeSets.has(sets.cursorHoldingInteractable)) return INVALID;
+    if (userinput.get(paths.actions.cursor.scaleGrabbedGrabbable)) return FINISH;
+    return VALID;
+  },
+  pen_color: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.get(paths.actions.cursor.penNextColor) || userinput.get(paths.actions.cursor.penPrevColor)) {
+      return FINISH;
+    }
+    return VALID;
+  },
+  pen_size: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.get(paths.actions.cursor.scalePenTip)) return FINISH;
+    return VALID;
+  },
+  pen_mode: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    return VALID;
+  },
+  freeze_mode: function(userinput, scene) {
+    return scene.is("frozen") ? VALID : INVALID;
+  },
+  video_share_mode: function(userinput, scene) {
+    return scene.is("sharing_video") ? VALID : INVALID;
+  },
+  mirror_mode: function(userinput, scene) {
+    return scene.is("mirroring") ? VALID : INVALID;
+  },
+  mute_mode: function(userinput, scene) {
+    return scene.is("muted") ? VALID : INVALID;
+  }
+};
+
+AFRAME.registerSystem("tips", {
+  init: function() {
+    this.activeTips = {};
+    this._finishedScopes = {};
+    this._performStep = this._performStep.bind(this);
+
+    if (localStorage.getItem(LOCAL_STORAGE_KEY) === null) {
+      localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify({}));
+    }
+
+    this._tickCount = 0;
+  },
+
+  tick: function() {
+    this._tickCount++;
+
+    if (!this._userinput) {
+      this._userinput = AFRAME.scenes[0].systems.userinput;
+
+      if (!this._userinput) return;
+    }
+
+    if (!this._mediaCounter) {
+      this._mediaCounter = document.querySelector("#media-counter").components["networked-counter"];
+
+      if (!this._mediaCounter) return;
+    }
+
+    const tips = platformTips();
+
+    const prevTop = this.activeTips.top;
+    const prevBottom = this.activeTips.bottom;
+
+    this._performStep(tips.top, "top");
+    this._performStep(tips.bottom, "bottom");
+
+    if (prevTop !== this.activeTips.top || prevBottom !== this.activeTips.bottom) {
+      this.el.emit("tips_changed", this.activeTips);
+    }
+  },
+
+  _performStep: function(tips, scope) {
+    if (finishedScopes[scope]) return;
+
+    const scene = AFRAME.scenes[0];
+    let chosenTip = null;
+    let finishCount = 0;
+
+    for (let i = 0; i < tips.length; i++) {
+      const tip = tips[i];
+      if (isTipFinished(tip)) {
+        finishCount++;
+        continue;
+      }
+
+      switch (VALIDATORS[tip](this._userinput, scene, this._mediaCounter, window.APP.store)) {
+        case FINISH:
+          markTipFinished(tip);
+          break;
+        case VALID:
+          chosenTip = `${tipPlatform()}.${tip}`;
+          break;
+      }
+
+      if (chosenTip) break;
```

Comment:
Hmm. If I finished a tip this frame (even if I wasnt seeing the tip), seems like we still want to mark it as finished. This early out might skip some.

Reply:
that's a tough one actually. i think i'm going to leave it as is for now until we see if its annoying without doing this. if I did this, then there's probably some flows of FTUE where the user incidentally hits some feature without really grokking it and we blow past the tip. (eg, if they accidentally hit Q or something, they'll never actually be told Q's functionality.)

Reply:
> they'll never actually be told Q's functionality
They had something even better, they experienced it, and now they don't need to be told.